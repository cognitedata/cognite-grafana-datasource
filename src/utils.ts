import { isNil, omitBy, get } from 'lodash';
import { stringify } from 'query-string';
import ms from 'ms';
import { QueryOptions, QueryTarget } from './types';
import { FilterTypes, ParsedFilter } from './parser/types';

export function getQueryString(obj: any) {
  return stringify(omitBy(obj, isNil));
}

export function toGranularityWithLowerBound(milliseconds: number, lowerBound = 1000): string {
  return ms(Math.max(milliseconds, lowerBound));
}

// used for generating the options.requestId
export function getRequestId(options: QueryOptions, target: QueryTarget) {
  return `${options.dashboardId}_${options.panelId}_${target.refId}`;
}

export const applyFilters = <T>(objs: T[], filters: ParsedFilter[]): T[] => {
  if (!filters.length) {
    return objs;
  }

  return objs.filter((obj) => filters.every((filter) => checkFilter(obj, filter)));
};

export const checkFilter = <T>(obj: T, { path, filter, value }: ParsedFilter): boolean => {
  const valueToFilter = get(obj, path, null);
  const regex = new RegExp(`^${value}$`);

  switch (filter) {
    case FilterTypes.RegexEquals:
      return regex.test(valueToFilter);
    case FilterTypes.RegexNotEquals:
      return !regex.test(valueToFilter);
    case FilterTypes.NotEquals:
      return value !== valueToFilter;
    default:
      return false;
  }
};
