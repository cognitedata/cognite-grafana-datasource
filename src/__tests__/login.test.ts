import { getMockedDataSource } from './utils';

jest.mock('@grafana/runtime');
type Mock = jest.Mock;

describe('Login with API key', () => {
  const ds = getMockedDataSource();
  const { backendSrv } = ds;

  function makeLoginResponse(loggedIn: boolean, project: string) {
    return Promise.resolve({
      data: {
        data: { loggedIn, project },
      },
      status: 200,
    });
  }

  describe('When given valid login info and correct project', () => {
    const response = makeLoginResponse(true, 'TestProject');

    it('should log the user in', async () => {
      backendSrv.datasourceRequest = jest.fn().mockReturnValue(response);
      const result = await ds.testDatasource();

      expect(backendSrv.datasourceRequest).toBeCalledTimes(1);
      expect((backendSrv.datasourceRequest as Mock).mock.calls[0][0]).toMatchSnapshot();
      expect(result).toMatchSnapshot();
    });
  });

  describe('When given valid login info but incorrect project', () => {
    const response = makeLoginResponse(true, 'WrongProject');

    it('should display an error message', async () => {
      backendSrv.datasourceRequest = jest.fn().mockReturnValue(response);
      const result = await ds.testDatasource();

      expect(backendSrv.datasourceRequest).toBeCalledTimes(1);
      expect((backendSrv.datasourceRequest as Mock).mock.calls[0][0]).toMatchSnapshot();
      expect(result).toMatchSnapshot();
    });
  });

  describe('When given invalid login info', () => {
    const response = makeLoginResponse(false, 'string');

    it('should display an error message', async () => {
      backendSrv.datasourceRequest = jest.fn().mockReturnValue(response);
      const result = await ds.testDatasource();

      expect(backendSrv.datasourceRequest).toBeCalledTimes(1);
      expect((backendSrv.datasourceRequest as Mock).mock.calls[0][0]).toMatchSnapshot();
      expect(result).toMatchSnapshot();
    });
  });
});

describe('Login with OAuth2', () => {
  const ds = getMockedDataSource({ oauthPassThru: true });
  const { backendSrv } = ds;

  function makeLoginResponse(loggedIn: boolean, project: string) {
    return Promise.resolve({
      data: {
        projects: [{ projectUrlName: project }],
      },
      status: 200,
    });
  }

  describe('When given valid login info and correct project', () => {
    const response = makeLoginResponse(true, 'TestProject');
    let result;

    beforeAll(async () => {
      backendSrv.datasourceRequest = jest.fn().mockReturnValue(response);
      result = await ds.testDatasource();
    });

    it('should send a correct request', async () => {
      expect(backendSrv.datasourceRequest).toBeCalledTimes(1);
      expect((backendSrv.datasourceRequest as Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it('should log the user in', async () => {
      expect(result).toMatchSnapshot();
    });
  });

  describe('When given valid login info but incorrect project', () => {
    const response = makeLoginResponse(true, 'WrongProject');

    it('should display an error message', async () => {
      backendSrv.datasourceRequest = jest.fn().mockReturnValue(response);
      expect(await ds.testDatasource()).toMatchSnapshot();
    });
  });

  describe('When given invalid login info', () => {
    const response = makeLoginResponse(false, 'string');

    it('should display an error message', async () => {
      backendSrv.datasourceRequest = jest.fn().mockReturnValue(response);
      expect(await ds.testDatasource()).toMatchSnapshot();
    });
  });
});
