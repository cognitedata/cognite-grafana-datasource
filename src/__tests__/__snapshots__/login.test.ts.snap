// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login When given invalid login info should display an error message 1`] = `
Object {
  "method": "GET",
  "url": "/api/datasources/proxy/6/cdf-api-key/login/status",
}
`;

exports[`Login When given invalid login info should display an error message 2`] = `
Object {
  "message": "Your Cognite credentials are invalid",
  "status": "error",
  "title": "Error",
}
`;

exports[`Login When given valid login info and correct project should log the user in 1`] = `
Object {
  "method": "GET",
  "url": "/api/datasources/proxy/6/cdf-api-key/login/status",
}
`;

exports[`Login When given valid login info and correct project should log the user in 2`] = `
Object {
  "message": "Your Cognite credentials are valid",
  "status": "success",
  "title": "Success",
}
`;

exports[`Login When given valid login info but incorrect project should display an error message 1`] = `
Object {
  "method": "GET",
  "url": "/api/datasources/proxy/6/cdf-api-key/login/status",
}
`;

exports[`Login When given valid login info but incorrect project should display an error message 2`] = `
Object {
  "message": "Your Cognite credentials are invalid",
  "status": "error",
  "title": "Error",
}
`;
